### DISCUSSION TITLE
Unable to add MGT Person component to extensible library

### DISCUSSION THREAD

**Discussion by jamesderrick** [microsoft-search/pnp-modern-search#3447] [Q&A] [answered]
Hi, struggling to find a solution for this one.
I have created an extension library for the search results web part, to allow me to use react to create a custom layout
Everything works fine until I try and add the MGT Person component
Initially I imported the component as follows
`import { Person, ViewType } from '@microsoft/mgt-react';`
And used it like this
```
<div className="newsSubTitle">
   <Person personQuery={CreatedBy} view={ViewType.oneline}></Person>
</div>
```
But that was giving me lots of errors about the all the mgt components already being defined
So after some digging I found advice to import the components like this
```
import { Person } from '@microsoft/mgt-react/dist/es6/spfx';
import { ViewType } from '@microsoft/mgt-spfx';
``` 
However, when I do that the library is no longer recognised by modern search
`Could not load test-library in require. TypeError: Cannot read properties of undefined (reading 'id')`

Any help would be really appreciated

**Comment by kasperbolarsen** [comment]
Hi James, 👋 

I am not capable to answer the extention question,  but I have tweaked the People layout a lot lately.  Are you sure that you can't achieve your goals with a custom layout?

**Reply by jamesderrick** [reply]
Hi Kasper, I did originally have it working as a custom layout, but would really like to be able to do it as an extension.
So it's not the end of the world if I can't solve my issue, but I feel like I'm very close to sorting it, if not for the issue mentioned above.

**Reply by jamesderrick** [reply]
Starting to think it might be a bug with pnp-modern-search
I can get the mgt person component to work fine with spfx without react, declaring the component like this...
`<mgt-person-card person-query="me"></mgt-person-card>`
And I get it to work fine with spfx and using the react component with imports
```
import { Person } from '@microsoft/mgt-react/dist/es6/spfx';
import { ViewType } from '@microsoft/mgt-spfx';
```
And using the component like this
`<Person personQuery='me' view={ViewType.oneline} />`
So the only scenario that doesn't work, is when I try and do it as part of an extension library

**Reply by wobba** [reply]
You don't need to import mgt-spfx in your custom component as that is loaded with the main web part. You can use mgt-person-card by using `<pnp-persona>` which wraps it. If you want to reference yourself you need to add code like we have to check if components are already registered.

https://github.com/microsoft-search/pnp-modern-search/blob/main/search-parts/src/helpers/GraphToolKitHelper.ts

**Reply by jamesderrick** [reply]
Thanks for the reply, but how would I use pnp-persona in my react component?
I get an error `Property 'pnp-persona' does not exist on type 'JSX.IntrinsicElements'` if I just try and add it into my render method
This is why I thought I needed to use the react components from mgt-react

**Reply by wobba** [reply, marked as answer]
You can emit it as markup, as it will be a valid tag on the page like any other html tag, not as something you consume. If you want to get the react component follow something similar to the file I point to for loading and referencing the toolkit. Multiple solutions cannot load and bootstrap the toolkit in the same DOM.

**Reply by jamesderrick** [reply]
Thanks, don't think I necessarily did it the way you recommended but I now have it working, feels like a bit of a hack but progress nonetheless
```
const authorInfo = CreatedBy.split('|');
const reactString = `<mgt-person class="person" person-query="${authorInfo[0].trim()}" view="oneline"></mgt-person>`;
const personElement = React.createElement('div', { key: 'person-' + DocId, dangerouslySetInnerHTML: { __html: reactString}});
```
And then I output the new element in the render method

**Reply by wobba** [reply]
That works.. you re-use the already loaded mgt-person. What extra logic are you adding since you create a custom extension instead of using the pnp-persona component in a handlebar template directly?

**Reply by jamesderrick** [reply]
We just have a lot of templates that are very similar, almost identical, so I'm using the extension functionality to allow us to just have the one template, but with a fair few layout properties. Seemed easier to me to do that with react components.
